### PR TITLE
Integrate the CubeArtisan draftbots.

### PR DESCRIPTION
**Pull request by ruler501**
In tests on CubeCobra data these bots had 60/81/90% accuracy for top 1/2/3 picks. They also appear to be quite fast, but I haven't benchmarked it yet.

I don't know enough about the codebase to appropriately assess the impact of these functions moving to async. It also increases the size of the card data file slightly since it now needs to store oracle ids as well as card ids since that's what the draftbots use so they don't have to worry about printings.

Further work can be done to get them working on the other draft formats supported (except maybe winston which has such a fundamentally different structure) if there is interest. The draftbots can handle any situation where they are picking between options of different combinations of cards (such as grid draft choosing rows or columns).

### CODE REVIEW

**Comment by Senryoku** [comment]
I'm having trouble running this locally, the thread library fails to load your worker on startup. Patching in the path without the leading "\\" solves the problem, `rebaseScriptPath` in thread.js may have some problem handling Windows?

```
internal/modules/cjs/loader.js:880
  const err = new Error(message);
              ^

Error: Cannot find module '\\H:\Source\MTGADraft_PRs\node_modules\mtgdraftbots\node\mtgdraftbotsWorker.js'
Require stack:
- H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.node.js
- H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.js
- H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.resolve (internal/modules/cjs/helpers.js:94:19)
    at resolveScriptPath (H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.node.js:84:27)
    at new Worker (H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.node.js:97:19)
    at file:///H:/Source/MTGADraft_PRs/node_modules/mtgdraftbots/node/mtgdraftbots.js:3:25
    at ModuleJob.run (internal/modules/esm/module_job.js:146:23)
    at async Loader.import (internal/modules/esm/loader.js:165:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'H:\\Source\\MTGADraft_PRs\\node_modules\\threads\\dist\\master\\implementation.node.js',
    'H:\\Source\\MTGADraft_PRs\\node_modules\\threads\\dist\\master\\implementation.js',
    'H:\\Source\\MTGADraft_PRs\\node_modules\\threads\\dist\\master\\index.js'
  ]
}
```

While I try to solve this correctly, I still have some questions!
You mention in the readme having to retrieve parameters from a remote server. 
 - Could you provide a separate method to do it ahead of time, rather than on the first call to calculateBotPick? This could also help checking for errors and having a backup bot implementation (reverting to our previous, simpler implementation).
 - Alternatively/Additionally, would it be possible to include a default version of these parameters that could be checked in the repo? (Similar to what I'm doing with the card data in the python script). I'd rather not be dependent on a remote server, even if I have to manually update these parameters from time to time.
 - What is that server btw? I wasn't able to find it in the source :O

Tests reports `(node:1767) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 message listeners added to [Worker]. Use emitter.setMaxListeners() to increase limit` and tests involving bots fail, not sure why yet as I haven't been able to reproduce the problem manually.

File: src/Bot.ts [hunk 1]
```diff
@@ -1,44 +1,67 @@
 "use strict";
 
+import { calculateBotPick } from "mtgdraftbots";
+
 import { Card } from "./Cards";
 
+const BASICS = [
```
Comment (Senryoku) [inline src/Bot.ts:7]:
```suggestion
const BasicsOracleIDs = [
```

File: src/Bot.ts [hunk 2]
```diff
@@ -1,44 +1,67 @@
...
-				maxScore = score;
-				bestPick = idx;
-			}
-		}
-		for (let color of booster[bestPick].colors) {
-			this.pickedColors[color] += 1;
-		}
+	getBotResult(booster: Card[], boosterNum: number, numBoosters: number, pickNum: number, numPicks: number) {
+		this.seen.push(...booster);
+		const cardOracleIds = BASICS.concat(
```
Comment (Senryoku) [inline src/Bot.ts:28]: Couldn't we incrementally build this array, instead of recreating it from scratch?

File: src/Bot.ts [hunk 3]
```diff
@@ -1,44 +1,67 @@
...
+			booster.map(c => c.oracle_id),
+			this.cards.map(c => c.oracle_id),
+			this.seen.map(c => c.oracle_id)
+		);
+		const drafterState = {
+			basics: [0, 1, 2, 3, 4],
+			cardsInPack: booster.map((_, idx) => idx + 5),
+			picked: this.cards.map((_, idx) => idx + 5 + booster.length),
```
Comment (Senryoku) [inline src/Bot.ts:36]: Related to the previous comment, and if I understand this correctly, by appending the booster oracle ids to the end of `cardOracleIds` we can skip this 'translation' and built this array incrementally too (`this.cards` can simply store `oracle_id`'s directly)

Comment (ruler501) [reply src/Bot.ts:36]: Yeah that's a great idea, I'll make the changes to do it. It'll also allow us to save some computation by using the same indices for booster as the cards we add to seen. Could also work to dedupe the array if you care about the small improvements there.

**Comment by ruler501** [comment]
> I'm having trouble running this locally, the thread library fails to load your worker on startup. Patching in the path without the leading "\" solves the problem, `rebaseScriptPath` in thread.js may have some problem handling Windows?
> ```
> internal/modules/cjs/loader.js:880
>   const err = new Error(message);
>               ^
> 
> Error: Cannot find module '\\H:\Source\MTGADraft_PRs\node_modules\mtgdraftbots\node\mtgdraftbotsWorker.js'
> Require stack:
> - H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.node.js
> - H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.js
> - H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\index.js
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
>     at Function.resolve (internal/modules/cjs/helpers.js:94:19)
>     at resolveScriptPath (H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.node.js:84:27)
>     at new Worker (H:\Source\MTGADraft_PRs\node_modules\threads\dist\master\implementation.node.js:97:19)
>     at file:///H:/Source/MTGADraft_PRs/node_modules/mtgdraftbots/node/mtgdraftbots.js:3:25
>     at ModuleJob.run (internal/modules/esm/module_job.js:146:23)
>     at async Loader.import (internal/modules/esm/loader.js:165:24)
>     at async Object.loadESM (internal/process/esm_loader.js:68:5) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [
>     'H:\\Source\\MTGADraft_PRs\\node_modules\\threads\\dist\\master\\implementation.node.js',
>     'H:\\Source\\MTGADraft_PRs\\node_modules\\threads\\dist\\master\\implementation.js',
>     'H:\\Source\\MTGADraft_PRs\\node_modules\\threads\\dist\\master\\index.js'
>   ]
> }
> ```

I noticed that before but thought I'd fixed it. I primarily tested on Linux though. If you find a patch I'd gladly accept it. 

> While I try to solve this correctly, I still have some questions!
> You mention in the readme having to retrieve parameters from a remote server.
> 
> * Could you provide a separate method to do it ahead of time, rather than on the first call to calculateBotPick? This could also help checking for errors and having a backup bot implementation (reverting to our previous, simpler implementation).
> * Alternatively/Additionally, would it be possible to include a default version of these parameters that could be checked in the repo? (Similar to what I'm doing with the card data in the python script). I'd rather not be dependent on a remote server, even if I have to manually update these parameters from time to time.
> * What is that server btw? I wasn't able to find it in the source :O

The server is just Google cloud storage in the cubeartisan bucket. If you look at src/emscripten.cpp you'll find the link.

I'll look into a separate method, but it's complicated by it being in web assembly. Since I moved everything into a webworkers synchronous xhr might work. It does already have a default for everything to handle failed loads, but it is the same for every card/weight. The parameters themselves are larger than I'd be comfortable with distributing separately since it is intended to be usable from a browser as well (they're 3.8 MiB). It's also just difficult to pass from javascript to the web assembly since it's a custom binary format. How important is it to implement for you? 

> Tests reports `(node:1767) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 message listeners added to [Worker]. Use emitter.setMaxListeners() to increase limit` and tests involving bots fail, not sure why yet as I haven't been able to reproduce the problem manually.

This might be the test harness not supporting workers. It could also be issues with the threads library.

**Comment by Senryoku** [comment]
I deployed it to make sure it wasn't a problem with my local environment: Bots seems to always pick the first card of the pack.

> It's also just difficult to pass from javascript to the web assembly since it's a custom binary format. How important is it to implement for you?

Pretty important honestly, I'd very much prefer if we were not dependent on another service (or at least have a backup plan). You don't have to bundle the parameters with the library, providing an alternative method of initialization (ideally one that can load a local copy of the parameters) is enough. 
I'll have to tinker a little more with your library to understand the challenges you face, I'll try to get more familiar with it :)

**Comment by ruler501** [comment]
Make sure you aren't testing with afr cards. I don't have data for them so it isn't trained on them. I can check some things, but I definitely tested this library with cubeartisan in the browser. Maybe there's something I missed for it behaving in node. I'll look into it.

Would accepting a custom url for initialization work?

**Comment by ruler501** [comment]
I could also make a version with bundled parameters, as a separate package, but you'd need to update the package every time the bots were retrained if you wanted to stay up to date. Depending on interest and my money situation I'll try to retrain once a week or once a month. I think I can get them trained for about $10 a time if it doesn't need to be perfect.

If you'd like afr to work and have data from drafts with them I can add them to the training. Same with mh2. I'd just need the data that goes into the drafterstate object for each pick.

**Comment by Senryoku** [comment]
> Make sure you aren't testing with afr cards. I don't have data for them so it isn't trained on them. I can check some things, but I definitely tested this library with cubeartisan in the browser. Maybe there's something I missed for it behaving in node. I'll look into it.

Yes, I tried with multiple sets (you can try yourself here: https://mtgadraft-pr-draftbots-rt1yomq.herokuapp.com/ , check the picks in the draft logs afterwards).

> Would accepting a custom url for initialization work?

Yes! I haven't worked with emscripten in a while, so I'm not sure what's possible, but I see two possibilities:
 - Pass parameters via argc/argv to select an initialization method (e.g. no parameters for the current behavior, or a url/path to a file to load).
 - Do nothing in main and require a call to an initialization method before using the library.

> I could also make a version with bundled parameters, as a separate package, but you'd need to update the package every time the bots were retrained if you wanted to stay up to date. Depending on interest and my money situation I'll try to retrain once a week or once a month. I think I can get them trained for about $10 a time if it doesn't need to be perfect.

I don't want to force you to manage another package, I'm willing to retrieve the parameters manually myself and keep them up-to-date on my end, given I can provide a custom URL for the parameter file. (I already do that with the card data :) )

> If you'd like afr to work and have data from drafts with them I can add them to the training. Same with mh2. I'd just need the data that goes into the drafterstate object for each pick.

I stopped collecting logs and sadly don't have data for AFR or MH2, but if you have an endpoint that could accept them, it's pretty much ready (aside from converting it to a friendlier format, see https://github.com/Senryoku/MTGADraft/blob/master/src/Persistence.ts#L254).
That's another situation where we could fallback to our previous algorithm by the way, it would be interesting to have a way to detect that we don't have enough data for a specific set of cards.

**Comment by ruler501** [comment]
I'll probably add bindings to the initialization method. I can look into asyncify or synchronous xhr so the initialization can be awaited.

I'll look into the incorrect picks, how do I view the draft logs?

I could add a boolean array to the output that says what cards were recognized so it could fallback to another algorithm as needed.

**Comment by Senryoku** [comment]
> I'll look into the incorrect picks, how do I view the draft logs?

After you've completed a draft (you can set the timer to 1sec. for quick testing), it should appear under Game Logs on the top of the page. There you click on it to expand it and review each bot pick.

GIF 24-07-2021 02-47-24

**Comment by Senryoku** [comment]
> I could add a boolean array to the output that says what cards were recognized so it could fallback to another algorithm as needed.

I would be perfect if this was another method that could be called prior to calculateBotPick, so we'd just instantiate suitable bots at the start of the draft. It should take an array of oracle ids and return how many of them it recognize (or even a more sophisticated metric of how it would perform on this specific set of cards, based on the amount of data for each card for example, if you have that kind of information at the ready). The app will then choose to fallback or not based on that number.

File: src/Bot.ts [hunk 4]
```diff
@@ -1,44 +1,69 @@
...
-		}
+	getBotResult(booster: Card[], boosterNum: number, numBoosters: number, pickNum: number, numPicks: number) {
+		const boosterIdxs = booster.map((_, idx) => idx + this.oracleIds.length);
+		this.seen.push(...boosterIdxs);
+		this.oracleIds.push(...booster.map(({ oracle_id }) => oracle_id));
+		const drafterState = {
+			basics: [0, 1, 2, 3, 4],
+			cardsInPack: booster.map((_, idx) => idx + 5),
+			picked: this.cards.map((_, idx) => idx + 5 + booster.length),
+			seen: this.seen.map((_, idx) => idx + 5 + booster.length + this.cards.length),
```
Comment (Senryoku) [inline src/Bot.ts:37]:
```suggestion
			cardsInPack: boosterIdxs,
			picked: this.picked,
			seen: this.seen,
```
The five basics are already accounted for by `this.oracleIds.length`

File: src/Bot.ts [hunk 5]
```diff
@@ -1,44 +1,69 @@
...
+	"b34bb2dc-c1af-4d77-b0b3-a0fb342a5fc6",
+	"a3fb7228-e76b-4e96-a40e-20b5fed75685",
+];
+
 export default class Bot {
 	name: string;
 	id: string;
-	cards: Card[] = []; // For debugging mostly.
-	pickedColors: { [color: string]: number } = { W: 0, U: 0, R: 0, B: 0, G: 0 };
+	oracleIds: string[] = [];
```
Comment (Senryoku) [inline src/Bot.ts:18]:
```suggestion
	oracleIds: string[] = [...basicOracleIds];
```
Basics IDs have to be included, right?

**Review by Senryoku — COMMENTED** [review]
I think indices of oracleIds passed to calculateBotPick aren't correct anymore.

File: src/Bot.ts (hunk 4 above)
Comment (ruler501) [reply src/Bot.ts]: Yeah I forgot to include them and apparently to update the drafterstate object with the correct values when I updated everything else. Fixed now.

**Pull request closed without merging**